beginner_source/aws_distributed_training_tutorial.py

```diff
@@ -0,0 +1,675 @@
+# ~~~~~~~~~~~~~~~
+# 
+# To simplify the main loop, it is best to separate a training epoch step
+# into a function called ``train``. This function trains the input model
+# for one epoch of the *train\_loader*. The only distributed training
+# artifact in this function is setting the
+# `non\_blocking <https://pytorch.org/docs/stable/notes/cuda.html#use-pinned-memory-buffers>`__
+# attributes of the data and label tensors to ``True`` before the forward
+# pass. This allows asynchronous GPU copies of the data meaning transfers
+# can be overlapped with computation. This function also outputs training
+# statistics along the way so we can track progress throughout the epoch.
+# 
+# The other function to define here is ``adjust_learning_rate``, which
+# decays the initial learning rate at a fixed schedule. This is another
+# boilerplate trainer function that is useful to train accurate models.
+# 
+
+def train(train_loader, model, criterion, optimizer, epoch):
+
+    batch_time = AverageMeter()
+    data_time = AverageMeter()
+    losses = AverageMeter()
+    top1 = AverageMeter()
+    top5 = AverageMeter()
+
+    # switch to train mode
+    model.train()
+
+    end = time.time()
+    for i, (input, target) in enumerate(train_loader):
+
+        # measure data loading time
+        data_time.update(time.time() - end)
+
+        # Create non_blocking tensors for distributed training
+        input = input.cuda(non_blocking=True)
+        target = target.cuda(non_blocking=True)
+
+        # compute output
+        output = model(input)
+        loss = criterion(output, target)
+
+        # measure accuracy and record loss
+        prec1, prec5 = accuracy(output, target, topk=(1, 5))
+        losses.update(loss.item(), input.size(0))
+        top1.update(prec1[0], input.size(0))
+        top5.update(prec5[0], input.size(0))
+
+        # compute gradients in a backward pass
+        optimizer.zero_grad()
+        loss.backward()
+
+        # Call step of optimizer to update model params
+        optimizer.step()
+
+        # measure elapsed time
+        batch_time.update(time.time() - end)
+        end = time.time()
+
+        if i % 10 == 0:
+            print('Epoch: [{0}][{1}/{2}]\t'
+                  'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t'
+                  'Data {data_time.val:.3f} ({data_time.avg:.3f})\t'
+                  'Loss {loss.val:.4f} ({loss.avg:.4f})\t'
+                  'Prec@1 {top1.val:.3f} ({top1.avg:.3f})\t'
+                  'Prec@5 {top5.val:.3f} ({top5.avg:.3f})'.format(
+                   epoch, i, len(train_loader), batch_time=batch_time,
+                   data_time=data_time, loss=losses, top1=top1, top5=top5))
+
+def adjust_learning_rate(initial_lr, optimizer, epoch):
+    """Sets the learning rate to the initial LR decayed by 10 every 30 epochs"""
+    lr = initial_lr * (0.1 ** (epoch // 30))
+    for param_group in optimizer.param_groups:
+        param_group['lr'] = lr
+            
```

Comment:
This isn't enough to achieve the right accuracy, please refer to the learning schedule here.

https://github.com/pytorch/examples/pull/262/files